### PR TITLE
MS19925: Avatar Packager

### PR DESCRIPTION
**Pull request by thoys**
https://highfidelity.fogbugz.com/f/cases/19925/

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16112/HighFidelity-Beta-PR14632-ad2d9bc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16112/HighFidelity-Beta-Interface-PR14632-ad2d9bc.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16140/HighFidelity-Beta-PR14632-656d96c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16140/HighFidelity-Beta-Interface-PR14632-656d96c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16140/HighFidelity-Beta-PR14632-656d96c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16140/HighFidelity-Beta-Interface-PR14632-656d96c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16148/HighFidelity-Beta-PR14632-9e7f89e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16148/HighFidelity-Beta-Interface-PR14632-9e7f89e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16150/HighFidelity-Beta-PR14632-c21d037.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16150/HighFidelity-Beta-Interface-PR14632-c21d037.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16157/HighFidelity-Beta-PR14632-24aaeee.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16157/HighFidelity-Beta-Interface-PR14632-24aaeee.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16162/HighFidelity-Beta-PR14632-cb33a91.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16162/HighFidelity-Beta-Interface-PR14632-cb33a91.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16198/HighFidelity-Beta-PR14632-41effbf.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16198/HighFidelity-Beta-PR14632-41effbf.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16198/HighFidelity-Beta-Interface-PR14632-41effbf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16198/HighFidelity-Beta-PR14632-41effbf.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16198/HighFidelity-Beta-Interface-PR14632-41effbf.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16201/HighFidelity-Beta-PR14632-61ac9d0.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16201/HighFidelity-Beta-PR14632-61ac9d0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16201/HighFidelity-Beta-Interface-PR14632-61ac9d0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16211/HighFidelity-Beta-PR14632-9c96f7b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16211/HighFidelity-Beta-Interface-PR14632-9c96f7b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16211/HighFidelity-Beta-PR14632-9c96f7b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16211/HighFidelity-Beta-Interface-PR14632-9c96f7b.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16217/HighFidelity-Beta-PR14632-c13badc.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16217/HighFidelity-Beta-PR14632-c13badc.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16217/HighFidelity-Beta-Interface-PR14632-c13badc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16217/HighFidelity-Beta-PR14632-c13badc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16217/HighFidelity-Beta-Interface-PR14632-c13badc.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16223/HighFidelity-Beta-PR14632-c95da0a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16223/HighFidelity-Beta-PR14632-c95da0a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16223/HighFidelity-Beta-Interface-PR14632-c95da0a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16223/HighFidelity-Beta-PR14632-c95da0a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16223/HighFidelity-Beta-Interface-PR14632-c95da0a.exe

**Comment by jakram** [comment]
Currently testing

**Comment by jakram** [comment]
Paused testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16228/HighFidelity-Beta-PR14632-93f53cf.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16228/HighFidelity-Beta-PR14632-93f53cf.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16228/HighFidelity-Beta-Interface-PR14632-93f53cf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16228/HighFidelity-Beta-PR14632-93f53cf.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16228/HighFidelity-Beta-Interface-PR14632-93f53cf.exe

File: interface/resources/qml/hifi/avatarPackager/AvatarProject.qml [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    property int colorScheme;
+    property var uploader: undefined;
+
+    property bool hasSuccessfullyUploaded: true;
+
+    visible: false
+    anchors.fill: parent
+    anchors.margins: 10
+
+    function reset() { hasSuccessfullyUploaded = false }
```
Comment (dback2) [inline interface/resources/qml/hifi/avatarPackager/AvatarProject.qml]: should end in semicolon?

File: interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                    errorPopup.show("Error Writing Project File", "There was a problem while trying to write the FST file.");
+                    break;
+                case AvatarProjectStatus.ERROR_OPEN_INVALID_FILE_TYPE:
+                    errorPopup.show("Invalid Project Path", "The avatar packager can only open FST files.");
+                    break;
+                case AvatarProjectStatus.ERROR_OPEN_PROJECT_FOLDER:
+                    errorPopup.show("Project Missing", "Project folder cannot be found. Please locate the folder and copy/move it to its original location.");
+                    break;
+                case AvatarProjectStatus.ERROR_OPEN_FIND_FST:
+                    errorPopup.show("File Missing", "We cannot find the project file (avatar.fst) in the folder. Please locate it and move to the project folder.");
```
Comment (dback2) [inline interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml]: I would say cannot find it in the project folder

File: interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+                    break;
+                case AvatarProjectStatus.ERROR_OPEN_OPEN_FST:
+                    errorPopup.show("File Read Error", "We cannot read the project file (avatar.fst). Please make sure that it is not in use by another program.");
```
Comment (dback2) [inline interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml]: QFile.open returning false would only mean file is in use?

File: interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+        function displayErrorMessage(status) {
+            if (status === AvatarProjectStatus.SUCCESS) {
+                return;
+            }
+            switch (status) {
+                case AvatarProjectStatus.ERROR_CREATE_PROJECT_NAME:
+                    errorPopup.show("Project Folder Already Exists", "A folder with that name already exists at that location. Please choose a different project name or location.");
+                    break;
+                case AvatarProjectStatus.ERROR_CREATE_CREATING_DIRECTORIES:
+                    errorPopup.show("Project Folders Creation Error", "There was a problem during the creation of the Avatar Project directories. Please select a project location with write permissions.");
```
Comment (dback2) [inline interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml]: QDir.mkpath returning false would only mean does not have write permissions?

File: interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+                    break;
+                case AvatarProjectStatus.ERROR_CREATE_FIND_MODEL:
+                    errorPopup.show("Cannot Find Model File", "There was a problem while trying to find the specified model file. Please verify if it exist at the specified location.");
+                    break;
+                case AvatarProjectStatus.ERROR_CREATE_OPEN_MODEL:
+                    errorPopup.show("Cannot Open Model File", "There was a problem while trying to open the specified model file. Please verify if you have read permissions at the specified location.");
```
Comment (dback2) [inline interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml]: QFile.open returning false would only mean does not have read permissions?

File: interface/src/avatar/AvatarProject.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+QStringList AvatarProject::getScriptPaths(const QDir& scriptsDir) {
+    QStringList result{};
+    constexpr auto flags = QDir::Files | QDir::NoSymLinks | QDir::NoDotAndDotDot | QDir::Hidden;
+    if (!scriptsDir.exists()) {
+        return result;
+    }
+
+    for (auto& script : scriptsDir.entryInfoList({}, flags)) {
+        if (script.fileName().endsWith(".js")) {
```
Comment (dback2) [inline interface/src/avatar/AvatarProject.cpp]: Do we need a toLower on the extension check?

File: interface/src/avatar/AvatarProject.cpp [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+void AvatarProject::openInInventory() {
+    auto tablet = dynamic_cast<TabletProxy*>(
+        DependencyManager::get<TabletScriptingInterface>()->getTablet("com.highfidelity.interface.tablet.system"));
+    tablet->loadQMLSource("hifi/commerce/wallet/Wallet.qml");
+    DependencyManager::get<HMDScriptingInterface>()->openTablet();
+    auto name = getProjectName();
+
+    // I'm not a fan of this, but it's the only current option.
+    QTimer::singleShot(1000, [name]() {
```
Comment (dback2) [inline interface/src/avatar/AvatarProject.cpp]: magic number

File: interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml (hunk 4 above)
Comment (thoys) [reply interface/resources/qml/hifi/avatarPackager/AvatarPackagerApp.qml]: Hmm, do you have a better suggestion?

File: interface/resources/qml/hifi/avatarPackager/AvatarUploadStatusItem.qml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+Item {
+    id: root
+
+    height: 48
+
+    property string text: "NO STEP TEXT"
+    property int uploaderState;
+    property var uploader;
+
+    state: root.uploader === undefined ? "" :
```
[the remark this answers is not in the document]
Comment (thoys) [reply interface/resources/qml/hifi/avatarPackager/AvatarUploadStatusItem.qml]: Having them here like this makes it update the state automatically on any change of a variable used here. But I believe that states can also switch with the `when` condition property in each `State` object below there, maybe try that @huffman  ?

File: interface/src/avatar/AvatarProject.cpp [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        auto tablet = dynamic_cast<TabletProxy*>(
+            DependencyManager::get<TabletScriptingInterface>()->getTablet("com.highfidelity.interface.tablet.system"));
```
Comment (thoys) [inline interface/src/avatar/AvatarProject.cpp]: couldn't we re-use the same tablet object here that was already defined in `openInInventory()` ?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16241/HighFidelity-Beta-PR14632-b680e9c.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16250/HighFidelity-Beta-PR14632-51ce19e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16250/HighFidelity-Beta-Interface-PR14632-51ce19e.dmg

**Comment by jakram** [comment]
Will test (once built)

**Comment by huffman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16253/HighFidelity-Beta-PR14632-51ce19e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16253/HighFidelity-Beta-PR14632-51ce19e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16253/HighFidelity-Beta-Interface-PR14632-51ce19e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16253/HighFidelity-Beta-PR14632-51ce19e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16253/HighFidelity-Beta-Interface-PR14632-51ce19e.exe

**Comment by qlihifitest** [comment]
currently testing

**Comment by qlihifitest** [comment]
stopped testing -> Jamil is testing

File: libraries/networking/src/AccountManager.h [hunk 10]
```diff
@@ -148,15 +154,15 @@ private slots:
     QUrl _authURL;
 
     DataServerAccountInfo _accountInfo;
-    bool _isWaitingForTokenRefresh { false };
-    bool _isAgent { false };
+    bool _isWaitingForTokenRefresh{ false };
```
[the remark this answers is not in the document]
Comment (dback2) [reply libraries/networking/src/AccountManager.h]: This one still unaddressed

File: interface/src/avatar/AvatarPackager.h [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    AvatarPackager();
+    bool open();
+
+    Q_INVOKABLE AvatarProjectStatus::AvatarProjectStatus createAvatarProject(const QString& projectsFolder,
+                                                                             const QString& avatarProjectName,
+                                                                             const QString& avatarModelPath,
+                                                                             const QString& textureFolder);
+
+    Q_INVOKABLE AvatarProjectStatus::AvatarProjectStatus openAvatarProject(const QString& avatarProjectFSTPath);
+    Q_INVOKABLE bool isValidNewProjectName(const QString& projectPath, const QString& projectName) { return AvatarProject::isValidNewProjectName(projectPath, projectName); }
```
[the remark this answers is not in the document]
Comment (dback2) [reply interface/src/avatar/AvatarPackager.h]: This one still unaddressed

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16261/HighFidelity-Beta-PR14632-7433870.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16263/HighFidelity-Beta-PR14632-b0db81b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16263/HighFidelity-Beta-PR14632-b0db81b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16263/HighFidelity-Beta-Interface-PR14632-b0db81b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16263/HighFidelity-Beta-PR14632-b0db81b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16263/HighFidelity-Beta-Interface-PR14632-b0db81b.exe

**Pull request merged**